pkg/util/pki/certificatetemplate.go

```diff
@@ -0,0 +1,287 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki
+
+import (
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"time"
+
+	apiutil "github.com/cert-manager/cert-manager/pkg/api/util"
+	v1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	experimentalapi "github.com/cert-manager/cert-manager/pkg/apis/experimental/v1alpha1"
+	certificatesv1 "k8s.io/api/certificates/v1"
+)
+
+type CertificateTemplateMutator func(*x509.Certificate)
+
+// CertificateTemplateOverrideDuration returns a CertificateTemplateMutator that overrides the
+// certificate duration.
+func CertificateTemplateOverrideDuration(duration time.Duration) CertificateTemplateMutator {
+	return func(cert *x509.Certificate) {
+		cert.NotBefore = time.Now()
+		cert.NotAfter = cert.NotBefore.Add(duration)
+	}
+}
+
+// CertificateTemplateOverrideBasicConstraints returns a CertificateTemplateMutator that overrides
+// the certificate basic constraints.
+func CertificateTemplateOverrideBasicConstraints(isCA bool, maxPathLen *int) CertificateTemplateMutator {
+	return func(cert *x509.Certificate) {
+		cert.BasicConstraintsValid = true
+		cert.IsCA = isCA
+		if maxPathLen != nil {
+			cert.MaxPathLen = *maxPathLen
+			cert.MaxPathLenZero = *maxPathLen == 0
+		} else {
+			cert.MaxPathLen = 0
+			cert.MaxPathLenZero = false
+		}
+	}
+}
+
+// OverrideTemplateKeyUsages returns a CertificateTemplateMutator that overrides the
+// certificate key usages.
+func CertificateTemplateOverrideKeyUsages(keyUsage x509.KeyUsage, extKeyUsage []x509.ExtKeyUsage) CertificateTemplateMutator {
+	return func(cert *x509.Certificate) {
+		cert.KeyUsage = keyUsage
+		cert.ExtKeyUsage = extKeyUsage
+	}
+}
+
+// CertificateTemplateAddKeyUsages returns a CertificateTemplateMutator that adds the given key usages
+// to the certificate key usages.
+func CertificateTemplateAddKeyUsages(keyUsage x509.KeyUsage, extKeyUsage []x509.ExtKeyUsage) CertificateTemplateMutator {
+	return func(cert *x509.Certificate) {
+		cert.KeyUsage |= keyUsage
+
+	OuterLoop:
+		for _, usage := range extKeyUsage {
+			for _, existingUsage := range cert.ExtKeyUsage {
+				if existingUsage == usage {
+					continue OuterLoop
+				}
+			}
+
+			cert.ExtKeyUsage = append(cert.ExtKeyUsage, usage)
+		}
+	}
+}
```

Comment:
suggestion: Since this doesn't seem to be used anywhere, would it be worth removing it for now? Currently it risks increasing our maintenance burden without us knowing there's a need for it!